### DISCUSSION TITLE
Use rtsp stream to download video

### DISCUSSION THREAD

**Discussion by RafaJurado** [kokorin/Jaffree#370] [Q&A] [answered]
Hi. I'm kinda new to Jaffree so maybe this question can be irrelevant, but I'm stuck. 

I'm using Jaffree in a project involving IP Cameras from Hikvision, but not directly. The way to communicate with the cams is through a middleware that acts as a rest server, receiving requests and sending the appropriate response depending on what you asked for, nothing magical here. Also, the middleware has a bunch of devices registered in order to group communications. This devices are NVR (network video recorders). 

I can stream live or recorded files without problem, but I also need to download files. The middleware offers a "way" to do this that's identical to a live streaming: it gives you an rtsp URI for download that has to be returned to the middleware again inside a json.  The response to this is a data stream that I'm using as an InputStream for Jaffree using PipeInput.pumpFrom.

The problem arrives with the output. A file is created sloooowly in the project folder structure. In fact, the download takes as much time as the video length. I've tried with multiple options for the output and the input, but nothing changes. So, what I'm doing wrong?.

Some tests I've done:

- Right now I'm using the -re argument. I know this parameter is using for streaming, but without it ffmpeg throws some errors:
```
[mpeg @ 0x55b4d7b79700] [warning] Could not find codec parameters for stream 1 (Audio: adpcm_g722 (g722), mono, s16): unspecified sample rate
[mp4 @ 0x55b4d7c13980] [error] muxer does not support non seekable output
[error] Could not write header for output file #0 (incorrect codec parameters ?): Invalid argument
```

My actual code:
```
try (InputStream inputStream = entity.getContent();
		 BufferedOutputStream outputStream =
					 new BufferedOutputStream(new FileOutputStream("myVideo.mp4"))
	) {
			FFmpeg.atPath()
					.addInput(PipeInput.pumpFrom(inputStream)
							.addArgument("-re")
							.addArguments("-err_detect","ignore_err")
							.addArguments("-acodec","g722")
							.addArguments("-c:v","h264")
					)
					.setOverwriteOutput(true)
					.addOutput(	PipeOutput.pumpTo(outputStream)
							//UrlOutput.toPath(pathToDst)
							.addArguments("-c", "copy")
							.addArguments("-rtsp_transport", "tcp")
							.addArguments("-r","12")
							.disableStream(StreamType.AUDIO)
							.addArguments("-preset", "ultrafast")
							.setDuration(5, TimeUnit.MINUTES)
							.setFormat("mp4")
					)
					.execute();
				outputStream.flush();
				return outputStream;

```

**Comment by kokorin** [comment, marked as answer]
@RafaJurado  `-re` option doesn't fix the error, it just hides it. 

1. ffmpeg is smart enough to detect codecs automatically, try 
2. if you have RTSP URL it better to provide this URL using UrlInput
3. As output it's better to use ChannelOutput (which is seekable, notice `muxer does not support non seekable output` message)
4. It is even better to store output to file on local disk (UrlOutput)
5. you use `-c copy` which is useful when you only need to change media container, but not codec, framerate or cut video.  `-c copy` is not compatible with `-r 12` `-preset ultrafast` and duration

From architecture perspective I think it's better to download data stream to local disk first, then to convert it as you need and store result on local disk too, and only after that upload result to wherever you want.

**Reply by RafaJurado** [reply]
Thanks for answering so fast! (and thank you for Jaffree also).

1. I've tried and it's not working with the audio codec.
2. In theory the RTSP uri I have it's not the correct one to use. The middleware asks for this RTSP uri and searches for the correct device and the files associated with it. Then the middleware sends the data stream from the NVR.
3. It's a change I have to do, but by now the file created is seekable without problem.
4. That's just what I'm doing now.

 I've changed a bit the code and now the file is created fast enough in server folders. I know it's not perfect yet, but its usable right now:

```
Path pathToDst = Paths.get("video.mp4");
try (InputStream inputStream = entity.getContent();
	 BufferedOutputStream outputStream =
				 new BufferedOutputStream(new FileOutputStream("myVideo.mp4"))
	) {
		FFmpeg.atPath()
				.addInput(PipeInput.pumpFrom(inputStream)
							.addArguments("-err_detect","ignore_err")
							.addArguments("-acodec","g722")
							.addArguments("-c:v","h264")
				)
				.setOverwriteOutput(true)
				.addOutput(
						PipeOutput.pumpTo(outputStream)
							.addArguments("-c:v", "copy")
							.addArguments("-rtsp_transport", "tcp")
							.addArguments("-b:v","1024k")
							.disableStream(StreamType.AUDIO)
							.disableStream(StreamType.DATA)
							.setDuration(5, TimeUnit.MINUTES)
							.setFormat("ismv")
				)
				.execute();
				outputStream.flush();
	}
```

I'm sharing the code just in case it can be useful to someone.

> From architecture perspective I think it's better to download data stream to local disk first, then to convert it as you need and store result on local disk too, and only after that upload result to wherever you want.

Yes, you're right. That's what I'm trying to do now, but I have some problems using ProgressListener (complete noob with reactive programming in Java, I've worked with Kotlin and Swift reactive features, but not Java). I suppose I'll open a new thread if I'm not able to reach a solution.
